### PR TITLE
[FLINK-4856][state] Add MapState in KeyedState

### PR DESCRIPTION
**Pull request by shixiaogang**
1. Add `MapState` and `MapStateDescriptor`
2. Implementation of `MapState` in `HeapKeyedStateBackend` and `RocksDBKeyedStateBackend`.
3. Add accessors to `MapState` in `RuntimeContext`

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * <p>The state is only accessible by functions applied on a KeyedDataStream. The key is
+ * automatically supplied by the system, so the function always sees the value mapped to the
+ * key of the current element. That way, the system can handle stream and state partitioning
+ * consistently together.
+ *
+ * @param <UK> Type of the keys in the state.
+ * @param <UV> Type of the values in the state.
+ */
+@PublicEvolving
+public interface MapState<UK, UV> extends AppendingState<Map<UK, UV>, Iterable<Map.Entry<UK, UV>>> {
```
Comment (shixiaogang) [inline flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java]: `MapState` provides the `add` method which puts a collection of key-value pairs into the state. Though the semantics may be a little different in existing `AppendingState`s, I think it's okay for `MapState` to be an `AppendingState` because the interface does not enforce any restriction on the modification of previous data.

Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java]: Exactly, the semantics are different from how they are described on `AppendingState` and from how it is used. `MapState` can still have a `add(Map<UK, UV> map)` method but I don't think it makes sense that it is an `AppendingState`. What are the use cases where a user is expected to have an `AppendingState<Map<K, V>, Map<K, V>>` instead of simply having a `MapState<K, V>`?

As I said before, the semantics of `MapState.add()` would be like `ValueState.update()` and `ValueState` is also not an `AppendingState`. I think a `MultiMapState` that would be like a `Map<K, Iterable<V>>` (and would have the semantics of a `ListState`) would be able to satisfy `AppendingState` but I'm not sure whether we would need the interface there.

Comment (shixiaogang) [inline flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java]: I agree that it's `MultiMapState`, instead of `MapState`, that is supposed to be an `AppendingState`.  

I will update the interface hierarchy, making `MapState` not an `AppendingState`.

Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java]: Thanks 👍

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+			return;
+		}
+		
+		for (Map.Entry<UK, UV> entry : map.entrySet()) {
+			put(entry.getKey(), entry.getValue());
+		}
+	}
+	
+	@Override
+	public void clear() {
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: I wonder if it would be more efficient to just drop and recreate the column family?

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java [hunk 3]
```diff
@@ -834,7 +836,7 @@ private void restoreKVStateData() throws IOException, RocksDBException {
 	}
 
 	@Override
-	protected <N, T> InternalValueState<N, T> createValueState(
+	public <N, T> InternalValueState<N, T> createValueState(
```
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: I think these should stay `private`/`protected`. It was mistake to make them `public` in the other state backends. Creating a state can always be achieved with `getPartitionedState()` and a `StateDescriptor`.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 2 above)
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: I think that doesn't work be because the iterator doesn't iterator over the whole column family, just over the entries of the current state key, which can have several user-key/value pairs.

Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: Yes that is true, my bad. Then this shortcut doesn't work of course.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+
+	/** Serializer for the keys and values */
+	private final TypeSerializer<UK> userKeySerializer;
+	private final TypeSerializer<UV> userValueSerializer;
+
+	/**
+	 * We disable writes to the write-ahead-log here. We can't have these in the base class
+	 * because JNI segfaults for some reason if they are.
+	 */
+	private final WriteOptions writeOptions;
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: To be honest, i have no idea why we can't put `writeOptions` in base class. We put it in `AbstractRocksDBState` and  do not come across any problem in our production environment. 

Maybe @aljoscha is more familiar with the problem.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+
+	@Override
+	public UV get(UK userKey) throws IOException {
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+			byte[] rawValueBytes = backend.db.get(columnFamily, rawKeyBytes);
+
+			return (rawValueBytes == null ? null : deserializeUserValue(rawValueBytes));
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while getting data from RocksDB.", e);
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: This is to be consistent with the behavior in other RocksDB states.  I agree with you that we should use a more specific exception. It also helps to reduce the stack printed in the cases of exception.  But I think we should do the changes in a different pull request because it involves the changes in other states. What do you think?

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+		}
+	}
+
+	@Override
+	public void put(UK userKey, UV userValue) throws IOException {
+		if (userValue == null) {
+			remove(userKey);
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: I remove those mappings whose values from RocksDB because currently we have no method to serialize the null value. Many type serializers do not support the serialization of `null` (e.g. IntSerializer).  It also follows `RocksDBValueState` which removes the data when the value is updated to `null`.

In current implementation, the `get(UK)` method can still get the correct result if we remove the mapping from RocksDB because we always return `null` when the key is found in the RocksDB. The only concern is that the result returned by `contains(UK)` and `size()` will be different from a Map which supports `null` values. 

Actually, I prefer not to remove the entry if its value is `null`. But at prior to that, I think a different pull request is urgently needed to let all type serializers support `null` values.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+			return userKey;
+		}
+
+		@Override
+		public UV getValue() {
+			if (deleted) {
+				return null;
+			} else {
+				if (userValue == null) {
+					userValue = deserializeUserValue(rawValueBytes);
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: Good point. I will update it as suggested.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+			if (cacheIndex > cacheEntries.size()) {
+				throw new IllegalStateException();
+			}
+
+			// Load cache entries only when the cache is empty and there still exist unread entries
+			if (cacheIndex < cacheEntries.size() || expired) {
+				return;
+			}
+
+			RocksIterator iterator = db.newIterator(columnFamily);
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: Yeah, it's much better. Will update it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java [hunk 9]
```diff
@@ -93,6 +95,18 @@ public DefaultKeyedStateStore(KeyedStateBackend<?> keyedStateBackend, ExecutionC
...
 
+	@Override
+	public <UK, UV> MapState<UK, UV> getMapState(MapStateDescriptor<UK, UV> stateProperties) {
+		requireNonNull(stateProperties, "The state properties must not be null");
+		try {
+			stateProperties.initializeSerializerUnlessSet(executionConfig);
+			MapState<UK, UV> originalState = getPartitionedState(stateProperties);
+			return new UserFacingMapState<>(originalState);
+		} catch (Exception e) {
+			throw new RuntimeException("Error while getting state", e);
```
Comment (shixiaogang) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java]: Currently, `KeyedStateStore#getState()` does not throw exception in its declaration. `RuntimeException` is the only exception that can be thrown. Since the modification to the interface will affect user code (users will have to deal with thrown exceptions), I am not sure it's okay to modify the function declaration in `KeyedStateStore`.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java (hunk 3 above)
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: It is mainly due to the unit tests in `KvStateRequestSerializerTest` which need the accessors to `InternalKvState`.  A better choice to use `getPartitionState()` to obtain a user-facing state and convert it to an internal state. What do you think?

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializerTest.java [hunk 10]
```diff
@@ -410,6 +415,124 @@ public void testDeserializeListTooShort2() throws Exception {
...
+		TypeSerializer<Long> userKeySerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> userValueSerializer = StringSerializer.INSTANCE;
+		mapState.setCurrentNamespace(VoidNamespace.INSTANCE);
+
+		// List
+		final int numElements = 10;
+
+		final Map<Long, String> expectedValues = new HashMap<>();
+		for (int i = 0; i < numElements; i++) {
+			final long value = ThreadLocalRandom.current().nextLong();
```
Comment (shixiaogang) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializerTest.java]: I prefer to use `ThreadLocalRandom.current()` which is also used in other tests in this file. Though it makes difficult to reproduce the case, it may help to find corner cases.

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializer.java [hunk 11]
```diff
@@ -484,6 +487,71 @@ public static Throwable deserializeServerFailure(ByteBuf buf) throws IOException
...
+				while (it.hasNext()) {
+					Map.Entry<UK, UV> entry = it.next();
+
+					keySerializer.serialize(entry.getKey(), dos);
+					valueSerializer.serialize(entry.getValue(), dos);
+				}
+
+				return dos.getCopyOfBuffer();
+			} else {
+				return null;
```
Comment (shixiaogang) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializer.java]: The function is unused now. I will delete it in the update.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 12]
```diff
@@ -0,0 +1,579 @@
...
+			count++;
+			iterator.next();
+		}
+
+		return count;
+	}
+
+	@Override
+	public Iterable<UK> keys() {
+		return new Iterable<UK>() {
```
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: The `Iterable` is used to iterate over the mappings under the current key. Though we can reuse a singleton object for different keys, it will make the implementation complicated.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 4 above)
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: Do you have a core dump file for this problem that you could with us? I would try to take a look into this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java (hunk 9 above)
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java]: What I meant was using a more specific subclass of `RuntimeException` (e.g. `IllegalAccess`,`IllegalState`, `IndexOutOfBounds`, etc.) that better reflects the error, but still can leave the method signature unchanged.

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializerTest.java (hunk 10 above)
Comment (StefanRRichter) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/netty/message/KvStateRequestSerializerTest.java]: I understand your reason. Maybe we could also just print the random seed for each run, so that in case of a test error, we can immediately reproduce it?

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 5 above)
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: A different PR is fine. In this case, could you create a JIRA so that we do not forget about this point?

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 6 above)
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: I think that this would be a weakness of the serializer, or that some data types simply would not allow for `null` values (`null` would be a programming error for them). My feeling is that this code should not have to work around this problem. Probably any type serializer could trivally support null by returning a byte[0].

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 12 above)
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: What I meant was not a global singleton, just a single iterable per state. Since this is just a minor optimization, I could also just live with the code as is.

**Comment by StefanRRichter** [comment]
I have one more point that I forgot in my initial review. It would be great if you could also provide a paragraph in the Flink documentation about the MapState. This would improve the accessibility and visibility of this nice feature a lot for all users.

**Comment by shixiaogang** [comment]
@StefanRRichter I have updated the pull request as suggested. Now the map serializer supports the serialization of null values.

**Comment by shixiaogang** [comment]
I have added the documentation for `MapState`. You may take a look to see if it's properly written.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 4 above)
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: It was there initially because it would core dump in the very first version of the RocksDB backend otherwise. Not sure if this would still persist now. Maybe it was a problem with the cleanup.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java (hunk 3 above)
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: Hmm I see, let's keep it as you did it then.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java [hunk 13]
```diff
@@ -382,11 +342,26 @@ private UV deserializeUserValue(byte[] rawValueBytes) {
...
-		
+
+		public void remove() {
+			deleted = true;
+			rawValueBytes = null;
+
+			try {
+				db.remove(columnFamily, writeOptions, rawKeyBytes);
+			} catch (RocksDBException e) {
+				throw new RuntimeException("Error while removing data from RocksDB.", e);
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: The intention of my comment about `RuntimeException` was not about changing the method signatures for throwing RocksDBExcepion. My suggestion was to only use a proper subclass of `RuntimeException`. We should avoid using `RuntimeException` directly, similar to how we should avoid throwing the class `Exception` directly. I know that there is some code in Flink that does not follow this, but I think it is better code style to stick with more appropriate subclasses.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 4 above)
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: In any case, the problem still persists that the native resource `WriteOptions` is never closed properly. It might be ok, because right now the task will typically end when is should be released, but it is still not completely clean and can produce leaks. I would be ok with merging this and creating another JIRA for this. What do you think @aljoscha ?

**Comment by StefanRRichter** [comment]
Besides the problem with the `WriteOption`, which I suggest to resolve in another PR, this looks good to merge for me now. +1 from me and waiting for the second approval through @aljoscha.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 4 above)
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: Yes, I think we need a follow-up issue because the other state types also have the `WriteOptions` that are not properly cleaned up.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java (hunk 13 above)
Comment (shixiaogang) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java]: I modify the method signature because I find, except `ValueState`, the methods in other states all throw `Exception`. I think it's okay because `MapState` is a common interface which has no idea of the implementation.  The implementation  of these methods, however, should throw some specific exception like `IOException` or `RocksDBException`. 

I think it's reasonable. What do you think?

**Comment by shixiaogang** [comment]
@StefanRRichter  Very thanks for your work.  I have rebased the pull request and resolved the conflicts.

File: flink-core/src/main/java/org/apache/flink/api/common/state/MapStateDescriptor.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.MapSerializer;
+import org.apache.flink.api.java.typeutils.MapTypeInfo;
+
+import java.util.Map;
+
+/**
+ * A {@link StateDescriptor} for {@link MapState}. This can be used to create state where the type
+ * is a list that can be appended and iterated over.
```
Comment (wenlong88) [inline flink-core/src/main/java/org/apache/flink/api/common/state/MapStateDescriptor.java]: commons error, caused by copy and paste

File: docs/dev/stream/state.md [hunk 15]
```diff
@@ -118,6 +118,11 @@ added to the state. Contrary to `ReducingState`, the aggregate type may be diffe
 of elements that are added to the state. The interface is the same as for `ListState` but elements
 added using `add(T)` are folded into an aggregate using a specified `FoldFunction`.
 
+* `MapState<UK, UV>`: This keeps a list of mappings. You can put key-value pairs into the state and retrieve
+retrieve an `Iterable` over all currently stored mappings. Mappings are added using `put(UK, UV)` or
```
Comment (aljoscha) [inline docs/dev/stream/state.md]: duplicate "retrieve"

File: docs/dev/stream/state.md [hunk 16]
```diff
@@ -118,6 +118,11 @@ added to the state. Contrary to `ReducingState`, the aggregate type may be diffe
 of elements that are added to the state. The interface is the same as for `ListState` but elements
 added using `add(T)` are folded into an aggregate using a specified `FoldFunction`.
 
+* `MapState<UK, UV>`: This keeps a list of mappings. You can put key-value pairs into the state and retrieve
+retrieve an `Iterable` over all currently stored mappings. Mappings are added using `put(UK, UV)` or 
+`putAll(map<UK, UV>)`. The value associated with a user key can be retrieved using `get(UK)`. The iterable
```
Comment (aljoscha) [inline docs/dev/stream/state.md]: lowercase `map`

**Comment by aljoscha** [comment]
I had to very minor comments about typos in Javadoc. Otherwise this looks very good now, @shixiaogang 👍 

@StefanRRichter Could you please go ahead and merge when you are satisfied with the internals?

**Comment by shixiaogang** [comment]
@aljoscha Thanks a lot for your hard work. I have fixed the typos in the documentation.

**Comment by aljoscha** [comment]
Thanks a lot for working on this @shixiaogang. ! 😃 

I just merged your PR, could you please close is.
